src/election_notification_works/handlers/signup/email.clj

```
@@ -15,56 +21,147 @@
    notification-works, where it will be merged with a template and sent out."
-  [send-fn user reg-form-and-info vbm-form-and-info mailings?]
+  [send-fn user electorate reg-form-and-info vbm-form-and-info mailings?]
   (log/debug "sending signup email for" (pr-str user))
   (let [reg-status (select-keys (reg-status/latest-reg-status (:id user))
                                 [:method :status])
         recipient (-> user
                       util/user->recipient
                       (assoc :election-mail-subscription mailings?
                              :registration-status reg-status))
+        state-ocd-id (->> (get-in recipient
+                                  [:addresses :registered :ocd-divisions])
+                          util/state-ocd-id)
+        state-authority (auth/get-state-authority state-ocd-id)
+        state-reg-method (matching-state-district-division-reg-method recipient state-authority)
+        reg-deadline-map (reg-deadline-and-type electorate)
         signers (util/signature)
         attachments (form-attachments reg-form-and-info vbm-form-and-info)
-        reg-form-ctx (->reg-form-ctx reg-form-and-info)
-        vbm-application-ctx (->vbm-application-ctx user vbm-form-and-info)
-        auth-lookup-urls (util/user->local-auth-lookup-urls user)]
+        vbm-application-ctx (or (->vbm-application-ctx user vbm-form-and-info)
+                                (default-vbm-ctx electorate user))
+        local-reg-authority (:authority (auth/local-reg-authority-search user))
+        state-authority-urls (util/user->state-authority-urls user)]
     (-> {:notification-type :turbovote-signup-confirmation
          :medium :email
          :language (:language-preference user)
-         :context (merge {:recipient recipient
-                          :signers signers}
-                         auth-lookup-urls
-                         reg-form-ctx vbm-application-ctx)
+         :context (remove-nil-vals
```

Do we need to remove nil vals? If it's important to remove them, is it also important to remove them recursively?

---

I suppose we don't really _need_ to. Typically I would take steps for each of several of the fns called to avoid merging the result into the context if the value is nil, but it seemed cleaner to eliminate them in one swell foop.

Offhand I don't remember seeing lower-level nils in this data, but I'll keep an eye out.

---

The one thing I'm (only slightly) concerned about with doing it at the end is that it might hide real errors. I agree that if there's a key with a `nil` value, that indicates a key that should have a value, and is an error. So it would serve as useful documentation if keys that should always be present could be found in the map that is the first argument to `merge` and keys that might not are in the rest of the maps.

---

Not a bad idea. I'll make a note to swing back around to this question on the refactor pass.